Train.cpp

```diff
@@ -537,69 +543,123 @@ int main(int argc, char** argv) {
           LOG(FATAL) << "Sample has NaN values - "
                      << join(",", readSampleIds(sample[kSampleIdx]));
         }
+        // This loop is required to ensure no training example is skipped while
+        // adjusting the loss scale factor. Note that when the gradient values
+        // are Inf/NaN, the results are skipped and the program tries to adjust
+        // the scale factor accordingly. The loop can help making the train
+        // process more deterministic. Also, it's reasonable to use these
+        // samples since they are already loaded in the memory hierarchy
+        // (ram/cache) and we are done computing their features.
+        bool retrySample = false;
+        do {
+          retrySample = false;
+
+          // forward
+          meters.fwdtimer.resume();
+          auto output = ntwrk->forward({fl::input(sample[kInputIdx])}).front();
+          output.cast(sample[kInputIdx].type());
+          af::sync();
+          meters.critfwdtimer.resume();
+          auto loss =
+              crit->forward({output, fl::noGrad(sample[kTargetIdx])}).front();
+          af::sync();
+          if (FLAGS_mixedprecision) {
+            ++scaleCounter;
+            loss = loss * scaleFactor;
+          }
+          meters.fwdtimer.stopAndIncUnit();
+          meters.critfwdtimer.stopAndIncUnit();
+
+          if (af::anyTrue<bool>(af::isNaN(loss.array())) ||
```

Comment:
This is an interesting idea for the scale factor and sample retrying, but I'm wondering — is this directly related to/required for AMP training in the experiments you did?

It's a significant change to the training flow, and I'd prefer if we moved the change to a separate PR so we can discuss in more detail if you were still able to train models with AMP without this scaling. That way we can be more deliberate about how best to integrate everything while ensuring good convergence.

Reply:
Skipping them is a little bit inefficient computation-wise since at this point the sample is already loaded and preprocessed. In addition, we will be skipping samples that triggered a gradient explosion. That is, samples that the model needs to learn better.